retrofit/src/main/java/retrofit2/RequestBuilder.java

```diff
@@ -107,19 +106,16 @@ private static String canonicalize(String input, boolean alreadyEncoded) {
     return input;
   }
 
-  private static void canonicalize(Buffer out, String input, int pos, int limit,
-      boolean alreadyEncoded) {
+  private static void canonicalizeForPath(Buffer out, String input, int pos, int limit,
+                                          boolean alreadyEncoded) {
     Buffer utf8Buffer = null; // Lazily allocated.
     int codePoint;
     for (int i = pos; i < limit; i += Character.charCount(codePoint)) {
       codePoint = input.codePointAt(i);
       if (alreadyEncoded
-          && (codePoint == '\t' || codePoint == '\n' || codePoint == '\f' || codePoint == '\r')) {
+          && shouldSkipCodePoint(codePoint)) {
         // Skip this character.
-      } else if (codePoint < 0x20
-          || codePoint >= 0x7f
-          || PATH_SEGMENT_ENCODE_SET.indexOf(codePoint) != -1
-          || (codePoint == '%' && !alreadyEncoded)) {
+      } else if (!alreadyEncoded && shouldEncodeCodePoint(codePoint)) {
```

Comment:
It is not re-encoded, but canonicalized as a value inside a single path
segment. With the current behavior, a slash is an impossible character to
exist inside a path segment without being transformed to an escaped version.

The path is currently treated as an ordered list of segments in both
Retrofit and OkHttp's HttpUrl class, which is what is leading to your
misinterpretation of the current behavior. It's not wrong, just explicit in
a way you don't agree with which results in surprise.

On Sun, Jan 31, 2016, 12:58 PM Jake Swenson <notifications@github.com>
wrote:

> In retrofit/src/main/java/retrofit2/RequestBuilder.java
> <https://github.com/square/retrofit/pull/1547#discussion_r51364383>:
>
> >          // Skip this character.
> > -      } else if (codePoint < 0x20
> > -          || codePoint >= 0x7f
> > -          || PATH_SEGMENT_ENCODE_SET.indexOf(codePoint) != -1
> > -          || (codePoint == '%' && !alreadyEncoded)) {
> > +      } else if (!alreadyEncoded && shouldEncodeCodePoint(codePoint)) {
>
> Thanks, but not sure if i understand. If the API creator says its encoded,
> why should if be re-encoded? If the caller passed in a bad value with the
> understanding it was to be encoded, why would they have any expectation it
> would work?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/square/retrofit/pull/1547/files#r51364383>.
>